File: lib/services/openweathermap_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:geocoding/geocoding.dart' as geocoding;
import '../models/weather_data.dart';
import '../models/location.dart';
import 'i_weather_service.dart';

class OpenWeatherMapService implements IWeatherService {
  static const String _baseUrl = 'https://api.openweathermap.org/data/2.5';
  static const String _geoUrl = 'https://api.openweathermap.org/geo/1.0';
  static const String _apiKey = '2f16c38d61c17ac94d944a5a66ca0e96';
  
  @override
  Future<WeatherData> getCurrentWeather(double lat, double lon) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/weather?lat=$lat&lon=$lon&appid=$_apiKey&units=metric'),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return _parseOpenWeatherMapCurrent(data);
      } else {
        throw Exception('Failed to load current weather from OpenWeatherMap');
      }
    } catch (e) {
      throw Exception('Error fetching current weather: $e');
    }
  }

  @override
  Future<WeatherData> getForecast(double lat, double lon, {int days = 7}) async {
    try {
      // OpenWeatherMap One Call API provides 7-day forecast
      final response = await http.get(
        Uri.parse('$_baseUrl/onecall?lat=$lat&lon=$lon&appid=$_apiKey&units=metric&exclude=minutely'),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return _parseOpenWeatherMapForecast(data);
      } else {
        throw Exception('Failed to load forecast from OpenWeatherMap');
      }
    } catch (e) {
      throw Exception('Error fetching forecast: $e');
    }
  }

  @override
  Future<List<WeatherData>> getHistoricalData(
    double lat, 
    double lon, 
    DateTime startDate, 
    DateTime endDate,
  ) async {
    // OpenWeatherMap historical data requires subscription
    throw Exception('Historical data requires OpenWeatherMap subscription');
  }

  @override
  Future<List<Location>> searchLocations(String query) async {
    try {
      final response = await http.get(
        Uri.parse('$_geoUrl/direct?q=$query&limit=5&appid=$_apiKey'),
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return data.map((json) => _parseOpenWeatherMapLocation(json)).toList();
      } else {
        // Fallback to geocoding service
        return await _searchUsingGeocoding(query);
      }
    } catch (e) {
      throw Exception('Error searching locations: $e');
    }
  }

  WeatherData _parseOpenWeatherMapCurrent(Map<String, dynamic> data) {
    final main = data['main'];
    final weather = data['weather'][0];
    final wind = data['wind'] ?? {};
    final sys = data['sys'] ?? {};
    
    return WeatherData(
      current: CurrentWeather(
        temperature: main['temp'].toDouble(),
        feelsLike: main['feels_like'].toDouble(),
        humidity: main['humidity'],
        pressure: main['pressure'].toDouble(),
        windSpeed: (wind['speed'] ?? 0).toDouble() * 3.6, // Convert m/s to km/h
        windDirection: wind['deg'] ?? 0,
        visibility: (data['visibility'] ?? 10000).toDouble() / 1000, // Convert m to km
        uvIndex: 0.0, // Not available in current weather endpoint
        condition: weather['description'] ?? '',
        icon: weather['icon'] ?? '',
        lastUpdated: DateTime.fromMillisecondsSinceEpoch(data['dt'] * 1000),
      ),
      daily: [], // Will be populated from forecast data
      hourly: [], // Will be populated from forecast data
      astronomy: Astronomy(
        sunrise: DateTime.fromMillisecondsSinceEpoch((sys['sunrise'] ?? 0) * 1000),
        sunset: DateTime.fromMillisecondsSinceEpoch((sys['sunset'] ?? 0) * 1000),
        moonPhase: 0.0, // Not available
        moonrise: DateTime.now(), // Not available
        moonset: DateTime.now(), // Not available
      ),
    );
  }

  WeatherData _parseOpenWeatherMapForecast(Map<String, dynamic> data) {
    // Parse the full forecast response from One Call API
    final current = data['current'];
    final dailyData = data['daily'] as List<dynamic>? ?? [];
    final hourlyData = data['hourly'] as List<dynamic>? ?? [];

    return WeatherData(
      current: CurrentWeather(
        temperature: current['temp'].toDouble(),
        feelsLike: current['feels_like'].toDouble(),
        humidity: current['humidity'],
        pressure: current['pressure'].toDouble(),
        windSpeed: current['wind_speed'].toDouble() * 3.6,
        windDirection: current['wind_deg'] ?? 0,
        visibility: (current['visibility'] ?? 10000).toDouble() / 1000,
        uvIndex: current['uvi'].toDouble(),
        condition: current['weather'][0]['description'],
        icon: current['weather'][0]['icon'],
        lastUpdated: DateTime.fromMillisecondsSinceEpoch(current['dt'] * 1000),
      ),
      daily: dailyData.take(7).map((day) => _parseDailyForecast(day)).toList(),
      hourly: hourlyData.take(24).map((hour) => _parseHourlyForecast(hour)).toList(),
      astronomy: _parseAstronomy(data),
    );
  }

  DailyForecast _parseDailyForecast(Map<String, dynamic> day) {
    final temp = day['temp'];
    final weather = day['weather'][0];
    
    return DailyForecast(
      date: DateTime.fromMillisecondsSinceEpoch(day['dt'] * 1000),
      maxTemp: temp['max'].toDouble(),
      minTemp: temp['min'].toDouble(),
      condition: weather['description'],
      icon: weather['icon'],
      chanceOfRain: ((day['pop'] ?? 0) * 100).round(),
    );
  }

  HourlyForecast _parseHourlyForecast(Map<String, dynamic> hour) {
    final weather = hour['weather'][0];
    
    return HourlyForecast(
      time: DateTime.fromMillisecondsSinceEpoch(hour['dt'] * 1000),
      temperature: hour['temp'].toDouble(),
      condition: weather['description'],
      icon: weather['icon'],
      chanceOfRain: ((hour['pop'] ?? 0) * 100).round(),
    );
  }

  Astronomy _parseAstronomy(Map<String, dynamic> data) {
    final daily = data['daily'][0];
    return Astronomy(
      sunrise: DateTime.fromMillisecondsSinceEpoch(daily['sunrise'] * 1000),
      sunset: DateTime.fromMillisecondsSinceEpoch(daily['sunset'] * 1000),
      moonPhase: daily['moon_phase'].toDouble(),
      moonrise: DateTime.fromMillisecondsSinceEpoch((daily['moonrise'] ?? daily['sunrise']) * 1000),
      moonset: DateTime.fromMillisecondsSinceEpoch((daily['moonset'] ?? daily['sunset']) * 1000),
    );
  }

  Location _parseOpenWeatherMapLocation(Map<String, dynamic> json) {
    return Location(
      name: json['name'],
      region: json['state'] ?? '',
      country: json['country'],
      latitude: json['lat'].toDouble(),
      longitude: json['lon'].toDouble(),
      timezone: 'UTC', // OpenWeatherMap doesn't provide timezone info
    );
  }

  Future<List<Location>> _searchUsingGeocoding(String query) async {
    try {
      final List<geocoding.Location> geocodingLocations = await geocoding.locationFromAddress(query);
      
      List<Location> results = [];
      
      for (var geocodingLocation in geocodingLocations.take(5)) {
        try {
          final List<geocoding.Placemark> placemarks = 
              await geocoding.placemarkFromCoordinates(geocodingLocation.latitude, geocodingLocation.longitude);
          
          if (placemarks.isNotEmpty) {
            final placemark = placemarks.first;
            final locationName = placemark.locality ?? 
                               placemark.subAdministrativeArea ?? 
                               placemark.administrativeArea ?? 
                               'Unknown';
            
            results.add(Location(
              name: locationName,
              region: placemark.administrativeArea ?? placemark.subAdministrativeArea ?? '',
              country: placemark.country ?? '',
              latitude: geocodingLocation.latitude,
              longitude: geocodingLocation.longitude,
              timezone: 'UTC',
            ));
          }
        } catch (e) {
          continue;
        }
      }
      
      return results;
    } catch (e) {
      throw Exception('Geocoding search failed: $e');
    }
  }

  @override
  WeatherApiProvider get provider => WeatherApiProvider.openWeatherMap;

  @override
  String get apiName => 'OpenWeatherMap';

  @override
  bool get isAvailable => true;
}

File: src/services/OpenWeatherMapService.ts
import axios from 'axios';
import { WeatherService, WeatherData, DailyForecast, HourlyForecast } from './types';

export class OpenWeatherMapService implements WeatherService {
  private apiKey: string;
  private readonly baseUrl = 'https://api.openweathermap.org/data/2.5';
  private readonly fallbackApiKey = '2f16c38d61c17ac94d944a5a66ca0e96'; // Demo key

  constructor(apiKey?: string) {
    this.apiKey = apiKey || this.fallbackApiKey;
  }

  async getForecast(lat: number, lon: number, days: number = 7): Promise<WeatherData> {
    try {
      const response = await axios.get(`${this.baseUrl}/forecast`, {
        params: {
          lat,
          lon,
          appid: this.apiKey,
          units: 'metric',
          cnt: Math.min(days * 8, 40) // 8 forecasts per day (3-hour intervals), max 40
        }
      });

      return this.transformForecastData(response.data);
    } catch (error) {
      console.error('Error fetching forecast from OpenWeatherMap:', error);
      throw new Error('Failed to fetch forecast data');
    }
  }

  isAvailable(): boolean {
    return this.apiKey.length > 0;
  }

  getApiSource(): string {
    return this.apiKey === this.fallbackApiKey ? 'OpenWeatherMap' : 'OpenWeatherMap (Custom)';
  }

  private transformForecastData(data: any): WeatherData {
    const hourlyForecast: HourlyForecast[] = data.list.slice(0, 24).map((item: any) => ({
      time: new Date(item.dt * 1000).toISOString(),
      temperature: item.main.temp,
      condition: item.weather[0].description,
      icon: `https://openweathermap.org/img/wn/${item.weather[0].icon}@2x.png`,
      humidity: item.main.humidity,
      windSpeed: item.wind.speed * 3.6,
      precipitationChance: (item.pop || 0) * 100,
      precipitationMm: item.rain?.['3h'] || item.snow?.['3h'] || 0,
      uvIndex: 0, // Not available in forecast endpoint
      pressure: item.main.pressure || 1013,
      visibility: 10 // Default visibility for OpenWeatherMap
    }));

    // Group by day for daily forecast
    const dailyGroups = data.list.reduce((groups: any, item: any) => {
      const date = new Date(item.dt * 1000).toDateString();
      if (!groups[date]) {
        groups[date] = [];
      }
      groups[date].push(item);
      return groups;
    }, {});

    const dailyForecast: DailyForecast[] = Object.entries(dailyGroups)
      .slice(0, 7)
      .map(([date, items]: [string, any]) => {
        const temps = items.map((item: any) => item.main.temp);
        const humidities = items.map((item: any) => item.main.humidity);
        const winds = items.map((item: any) => item.wind.speed * 3.6);
        const precipChances = items.map((item: any) => (item.pop || 0) * 100);
        const precipAmounts = items.map((item: any) => item.rain?.['3h'] || item.snow?.['3h'] || 0);

        return {
          date: new Date(date).toISOString().split('T')[0],
          maxTemp: Math.max(...temps),
          minTemp: Math.min(...temps),
          condition: items[0].weather[0].description,
          icon: `https://openweathermap.org/img/wn/${items[0].weather[0].icon}@2x.png`,
          humidity: Math.round(humidities.reduce((a: number, b: number) => a + b, 0) / humidities.length),
          windSpeed: Math.max(...winds),
          precipitationChance: Math.max(...precipChances),
          precipitationMm: precipAmounts.reduce((a: number, b: number) => a + b, 0)
        };
      });

    return {
      location: {
        name: data.city.name,
        country: data.city.country,
        lat: data.city.coord.lat,
        lon: data.city.coord.lon
      },
      current: {
        temperature: data.list[0].main.temp,
        condition: data.list[0].weather[0].description,
        icon: `https://openweathermap.org/img/wn/${data.list[0].weather[0].icon}@2x.png`,
        humidity: data.list[0].main.humidity,
        windSpeed: data.list[0].wind.speed * 3.6,
        windDirection: this.getWindDirection(data.list[0].wind.deg || 0),
        pressure: data.list[0].main.pressure,
        uvIndex: 0,
        visibility: 10, // Default value
        feelsLike: data.list[0].main.feels_like
      },
      forecast: {
        daily: dailyForecast,
        hourly: hourlyForecast
      },
      astronomy: {
        sunrise: new Date(data.city.sunrise * 1000).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
        sunset: new Date(data.city.sunset * 1000).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
        moonPhase: '', // OpenWeatherMap doesn't provide moon data
        moonIllumination: -1 // -1 indicates data not available
      }
    };
  }

  private getWindDirection(degrees: number): string {
    const directions = ['N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE', 'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW'];
    const index = Math.round(degrees / 22.5) % 16;
    return directions[index];
  }
}